contracts/Loan.sol

```diff
@@ -0,0 +1,550 @@
+    //   module.healthcheck() == LoanLib.STATUS.ACTIVE,
+    //   'Loan: no collateral to init'
+    // );
+    
+    for(uint i; i < modules.length; i++) {
+      require(IModule(modules[i]).init(), 'Loan: misconfigured module');
+    }
+
+    // probably also need to check that the modules have this Loan contract
+
+    // check spigot has control of revenue contracts here?
+    // transfer DEBT as origination fee?
+
+    // or can make initialized and have separate function for executing everything after agreed updon in mutualUpgrade.
+    // just annoying if one module doesnt work, both parties have to keep calling init() to ACTIVE loan instead of 
+    _updateLoanStatus(LoanLib.STATUS.ACTIVE);
+    return true;
+  }
+
+  /**
+   *  @dev - loops through all modules and returns their status if required last to savegas on override external calls
+   *        returns early if returns non-ACTIVE
+  */
+  function healthcheck() override external returns(LoanLib.STATUS status) {
+    if(principal + totalInterestAccrued > maxDebtValue)
+      return _updateLoanStatus(LoanLib.STATUS.OVERDRAWN);
+      
+    for(uint i; i < modules.length; i++) {
+       // should we differentiate failed calls vs bad statuses?
+      status = IModule(modules[i]).healthcheck();
+      if(status != LoanLib.STATUS.ACTIVE)
+        return _updateLoanStatus(status);
+    }
+    
+    return loanStatus;
+  }
+  
+  /**
+   * @dev - Loan borrower and proposed lender agree on terms
+            and add it to potential options for borrower to drawdown on
+            Lender and borrower must both call function for MutualUpgrade to add debt position to Loan
+   * @param amount - amount of `token` to initially deposit
+   * @param token - the token to be lent out
+   * @param lender - address that will manage debt position 
+  */
+  function addDebtPosition(
+    uint256 amount,
+    address token,
+    address lender
+  )
+    mutualUpgrade(lender, borrower) 
+    override external
+    returns(bool)
+  {
+    bool success = IERC20(token).transferFrom(
+      lender,
+      address(this),
+      amount
+    );
+    require(success, 'Loan: no tokens to lend');
+
+    positionIds += 1;
+    debts[positionIds] = DebtPosition({
+      lender: lender,
+      token: token,
+      principal: 0,
+      interestAccrued: 0,
+      deposit: amount
+    });
+
+    emit AddDebtPosition(lender, token, amount);
+
+    // also add interest rate model here?
+    return true;
+  }
+
+
+  //////////////////
+  // MAINTAINENCE //
+  //////////////////
+
+
+  /**
+    @notice see _accrueInterst()
+  */
+  function accrueInterest() override external returns(uint256) {
+    return _accrueInterest();
+  }
+
+
+
+  // Liquidation
+
+  function liquidate(
+    uint256 positionId,
+    uint256 amount,
+    address escrowToken
+  )
+    onlyArbiter
+    validPositionId(positionId)
+    external
+  {
+    require(loanStatus == LoanLib.STATUS.LIQUIDATABLE, "Loan: not liquidatable");
```

Comment:
@kibagateaux I think it would be good to do an update here first as the escrow contract could have been topped up since the last status update

Reply:
thats true. will add a separate ticket for it